crates/context/config/src/client/near.rs

```diff
@@ -0,0 +1,287 @@
+                    .await
+            }
+            Operation::Write { method } => {
+                network
+                    .mutate(contract_id, method.into_owned(), payload)
+                    .await
+            }
+        }
+    }
+}
+
+impl Network {
+    async fn query(
+        &self,
+        contract_id: AccountId,
+        method: String,
+        args: Vec<u8>,
+    ) -> Result<Vec<u8>, Error> {
+        let response = self
+            .client
+            .call(methods::query::RpcQueryRequest {
+                block_reference: BlockReference::latest(),
+                request: QueryRequest::CallFunction {
+                    account_id: contract_id,
+                    method_name: method,
+                    args: FunctionArgs::from(args),
+                },
+            })
+            .await
+            .map_err(|err| Error::Custom {
+                operation: "querying contract",
+                reason: err.to_string(),
+            })?;
+
+        match response.kind {
+            QueryResponseKind::CallResult(CallResult { result, logs: _ }) => Ok(result),
+            _ => Err(Error::InvalidResponse {
+                operation: "querying contract".to_owned(),
+            }),
+        }
+    }
+
+    async fn mutate(
+        &self,
+        contract_id: AccountId,
+        method: String,
+        args: Vec<u8>,
+    ) -> Result<Vec<u8>, Error> {
+        let response = self
+            .client
+            .call(methods::query::RpcQueryRequest {
+                block_reference: BlockReference::latest(),
+                request: QueryRequest::ViewAccessKey {
+                    account_id: self.account_id.clone(),
+                    public_key: self.secret_key.public_key().clone(),
+                },
+            })
+            .await
+            .map_err(|err| Error::Custom {
+                operation: "fetching account",
+                reason: err.to_string(),
+            })?;
+
+        let (nonce, permission, block_hash) = match response {
+            RpcQueryResponse {
+                kind: QueryResponseKind::AccessKey(AccessKeyView { nonce, permission }),
+                block_hash,
+                block_height: _,
+            } => (nonce, permission, block_hash),
+            _ => {
+                return Err(Error::InvalidResponse {
+                    operation: "fetching account".to_owned(),
+                })
+            }
+        };
+
+        if let AccessKeyPermissionView::FunctionCall {
+            allowance: _,
+            receiver_id,
+            method_names,
+        } = permission
+        {
+            if receiver_id != contract_id {
+                return Err(Error::Custom {
+                    operation: "mutating contract",
+                    reason: format!(
+                        "access key does not have permission to call contract: {}",
+                        contract_id
+                    ),
```

Comment:
It seems to me that a reason such as this is better formalised into an error enum variant...? Same with the others?

Reply:
These can and have been, the others.. not quite